controlsfx/src/main/java/impl/org/controlsfx/skin/NotificationBar.java

```diff
@@ -138,54 +137,83 @@ public NotificationBar() {
 
         // initialise close button area
         closeBtn = new Button();
-        closeBtn.setOnAction(new EventHandler<ActionEvent>() {
-            @Override public void handle(ActionEvent arg0) {
-                hide();
-            }
-        });
+        closeBtn.setFocusTraversable(false);
+        closeBtn.setOnAction(arg0 -> hide());
         closeBtn.getStyleClass().setAll("close-button"); //$NON-NLS-1$
         StackPane graphic = new StackPane();
         graphic.getStyleClass().setAll("graphic"); //$NON-NLS-1$
         closeBtn.setGraphic(graphic);
         closeBtn.setMinSize(17, 17);
         closeBtn.setPrefSize(17, 17);
-        closeBtn.setFocusTraversable(false);
         closeBtn.opacityProperty().bind(transition);
         GridPane.setMargin(closeBtn, new Insets(0, 0, 0, 8));
 
         GridPane.setValignment(closeBtn, VPos.TOP);
-        
+
         // put it all together
         updatePane();
+
+        focusedProperty().addListener((o, ov, hasFocus) -> {
+            if (hasFocus) {
+                actionsBar.requestFocus();
+            }
+        });
+
+        EventHandler<KeyEvent> keyPressEventHandler = event -> {
+            if (KeyCode.TAB.equals(event.getCode())) {
+                handleTraversal(event);
+            }
+            event.consume();
+        };
+        addEventHandler(KeyEvent.KEY_PRESSED, keyPressEventHandler);
     }
 
     void updatePane() {
         actionsBar = ActionUtils.createButtonBar(getActions());
         actionsBar.opacityProperty().bind(transition);
+        actionsBar.focusedProperty().addListener((o, ov, hasFocus) -> {
+            if (!actionsBar.getButtons().isEmpty()) {
+                actionsBar.getButtons().get(0).requestFocus();

```

Comment:
Do we care about thread-safety? If so, then it's possible that the button is removed between lines 175 and 176, resulting in a `NullPointerException`. Thoughts on an OOP alternative that upholds encapsulation? Maybe it's not possible, but something like:

```
actionBar.giveButtonFocus(OK_BUTTON);  // OK_BUTTON constant = 0
```

That would eliminate the need for the conditional (in this method anyway). The `actionsBar` class would need a method like:

```
public void giveButtonFocus( final int index ) {
  final var buttons = getButtons();

  // index is 0-based, size starts at 1 if not empty, so use >
  if( buttons.size() > index ) {
    buttons.get( index ).requestFocus();
  }
}
```

That's then reusable, thread-safe, and will throw neither an `NullPointerException` nor an `ArrayIndexOutOfBounds` exception.

Reply:
JavaFX is a single threaded GUI framework i.e. all scene-graph activities are done on the JavaFX application thread. Therefore, activities like removing of nodes, focus etc. are all inherently executed synchronously.